rllib/evaluation/worker_set.py

```diff
@@ -610,6 +664,30 @@ def valid_module(class_path):
 
         return worker
 
+    def _worker_health_check(self) -> List[int]:
+        """Performs a health-check on each remote worker.
+
+        Returns:
+            List of indices (into `self._remote_workers` list) of faulty workers.
+            Note that index=1 is the 0th item in `self._remote_workers`.
+        """
+        logger.info("Health checking all workers ...")
+        checks = []
+        for worker in self.remote_workers():
+            _, obj_ref = worker.sample_with_count.remote()
+            checks.append(obj_ref)
+
+        faulty_worker_indices = []
+        for i, obj_ref in enumerate(checks):
+            try:
+                ray.get(obj_ref)
+                logger.info("Worker {} looks healthy.".format(i + 1))
+            except RayError:
+                logger.exception("Worker {} is faulty.".format(i + 1))
+                faulty_worker_indices.append(i + 1)
```

Comment:
actually why do you + 1 here, and then -1 when you use the indices in remove_failed_workers and recreate_failed_workers?

Reply:
`i` starts from 0 (due to hos the for loop is executed with range), but remote worker indices start from 1.
We could make the for loop start from 1, though.

Reply:
hmm, ok.
I think you are only probing remote_workers() here, so it's 0th remote worker, 1th remote worker, etc.
and when you recreate them, you are also going through 0th entry in self.remote_worker, 1th entry, etc ...
so they seem to match exactly.

Reply:
Yup, but still: Remote worker's indices start from 1, not 0.

Here, we only use the i for:
* printing out, which remote worker indices are bad (ok)
* returning the list of faulty remote worker indices (which start from 1) (ok)

In the function that calls this method, we also interpret the return values as remote worker indices (starting from 1). Hence the `worker = self.remote_workers()[worker_index - 1]`.

I really think it's all ok :)

Reply:
I see. this is conceptually more correctly.